### DISCUSSION TITLE
how to use signals in triggering Activities

### DISCUSSION THREAD

**Discussion by vabs212** [elsa-workflows/elsa-core#1436] [Design & tech]
I am working on Signals. Can someone please provide me some info or tutorial on signals. Checked the document already but did not get what i was looking for  .
or please give me some info about the input variables used in signals .such as correlation ID and input object. and how to trigger the activity based on signals

**Comment by sfmskywalker** [comment]
Sure. To send input and correlation ID while sending a signal, you can do this:

```csharp
var input = "Some input".
var result = await _signaler.TriggerSignalAsync("my signal name", "Some input", correlationId: "some correlation ID");
```

For a more complete example with regards to triggering signals and dealing with input, this [blog post series](https://sipkeschoorstra.medium.com/building-workflow-driven-net-applications-with-elsa-2-part-1-44e08a9ba94b) might offer some helpful insights.

If you have a specific use case that is not covered by any of this nor the documentation, then please let me know what the specific requirements are.

**Reply by vabs212** [reply]
Thanks .but i want to know what does correelation ID stands for here and why do i need to provide it and what sort of input do i need to provide . Actually i have implemented a fork and then want to trigger one activity which sends mail when the parameter is set true so i need to check how to trigger that activity only not other one .

**Reply by sfmskywalker** [reply]
Correlation ID is just a value you can use to associate your workflows with. For example, it could be a customer ID, an employee ID, a document ID, or a random value. The purpose of correlating workflows with a value you control is so that you can later resume workflows by providing this correlation ID. 

Unless you have a workflow instance ID - in that case, you don't need a correlation ID.

When using signals, you can trigger them by specifying either a workflow instance ID or a correlation ID - both is unnecessary.

In your example with the fork and the send email activity, signals will only do you any good if you actually also use the **Signal Received** activity. That's the only activity that responds to signals.

If you provide a more complete description of what you are trying to do, I might be able to help you figure out what to do next.

**Reply by vabs212** [reply]
Sure . I have a scenario where user sends the input body such as {
"Name":"Vaibhav Sharma1",
"Email":"vaibhav.sharma@easyrewardz.com",
"Password":"India@5551",
"RepeatPassword":"India@5551",
"SendMail":true,
"SendSms":false
}

and on the basis of parameter snedmail and send SMS the subsequent activity will be callled. I have created a receive Signal to resume my actvity but getting  context.WorkflowExecutionContext.IsFirstPass as false and in one instance getting input as of my custom model and in other one getting as null >i am very much confused how to use receive signal for the above scnario.
Please help

**Reply by vabs212** [reply]
Here is my sample workflow json body 

{
  "$id": "1",
  "definitionId": "dcf755879b524b42a8f09b880f22f691",
  "versionId": "89967d7fa6b1462398dc03fb4254f7b7",
  "name": "UserRegistrationWorkFlow",
  "displayName": "UserRegistrationWorkFlow",
  "description": "UserRegistrationWorkFlow",
  "version": 2,
  "variables": {
    "$id": "2",
    "data": {}
  },
  "customAttributes": {
    "$id": "3",
    "data": {}
  },
  "contextOptions": {
    "$id": "4",
    "contextFidelity": "Burst"
  },
  "isSingleton": false,
  "persistenceBehavior": "Suspended",
  "deleteCompletedInstances": false,
  "isPublished": true,
  "isLatest": true,
  "activities": [
    {
      "$id": "5",
      "activityId": "1c871c36-64e3-4890-9b3d-5a749c0d6d8e",
      "type": "HttpEndpoint",
      "name": "HTTP Endpoint test",
      "displayName": "HTTP Endpoint",
      "persistWorkflow": true,
      "loadWorkflowContext": true,
      "saveWorkflowContext": true,
      "properties": [
        {
          "$id": "6",
          "name": "Path",
          "expressions": {
            "$id": "7",
            "Literal": "/test"
          }
        },
        {
          "$id": "8",
          "name": "Methods",
          "expressions": {
            "$id": "9",
            "Json": "[\"POST\"]"
          }
        },
        {
          "$id": "10",
          "name": "ReadContent",
          "expressions": {
            "$id": "11",
            "Literal": "true"
          }
        },
        {
          "$id": "12",
          "name": "TargetType",
          "expressions": {
            "$id": "13",
            "Literal": "RegistrationModel"
          }
        }
      ],
      "propertyStorageProviders": {}
    },
    {
      "$id": "14",
      "activityId": "191f0857-9f14-46b3-aaef-cb5f2d2a1735",
      "type": "CreateUser",
      "name": "Create User",
      "displayName": "Create User",
      "description": "Create User",
      "persistWorkflow": false,
      "loadWorkflowContext": true,
      "saveWorkflowContext": false,
      "properties": [],
      "propertyStorageProviders": {}
    },
    {
      "$id": "15",
      "activityId": "1e574748-7698-4106-b0ed-26cb5eefe402",
      "type": "Fork",
      "displayName": "Fork",
      "persistWorkflow": false,
      "loadWorkflowContext": false,
      "saveWorkflowContext": false,
      "properties": [
        {
          "$id": "16",
          "name": "Branches",
          "expressions": {
            "$id": "17",
            "Json": "[\"ValidateUsingEmail\",\"ValidateUsingOtp\"]"
          }
        }
      ],
      "propertyStorageProviders": {}
    },
    {
      "$id": "18",
      "activityId": "949fe4bf-1261-440b-a836-137a4f7ebd2a",
      "type": "ActivateUserEmail",
      "name": "Activate User Email",
      "displayName": "Activate User Email",
      "persistWorkflow": false,
      "loadWorkflowContext": false,
      "saveWorkflowContext": false,
      "properties": [],
      "propertyStorageProviders": {}
    },
    {
      "$id": "19",
      "activityId": "e5656747-01e2-4a6f-a9f5-f36fd9a2bfed",
      "type": "ActivateUserSms",
      "name": "Activate User Sms",
      "displayName": "Activate User Sms",
      "persistWorkflow": false,
      "loadWorkflowContext": false,
      "saveWorkflowContext": false,
      "properties": [],
      "propertyStorageProviders": {}
    },
    {
      "$id": "20",
      "activityId": "d06ae00f-2472-4b68-b757-47ebf92ad3d9",
      "type": "CustomReceiveTrigger",
      "name": "TriggerEmail",
      "displayName": "TriggerEmail",
      "description": "TriggerEmail",
      "persistWorkflow": false,
      "loadWorkflowContext": false,
      "saveWorkflowContext": false,
      "properties": [
        {
          "$id": "21",
          "name": "Signal",
          "expressions": {
            "$id": "22",
            "Literal": "EmailValidationSignal"
          }
        }
      ],
      "propertyStorageProviders": {}
    },
    {
      "$id": "23",
      "activityId": "b63178a5-d02c-4f0a-b19e-e7d42969277b",
      "type": "CustomReceiveTrigger",
      "name": "TriggerSms",
      "displayName": "TriggerSms",
      "description": "TriggerSms",
      "persistWorkflow": false,
      "loadWorkflowContext": false,
      "saveWorkflowContext": false,
      "properties": [
        {
          "$id": "24",
          "name": "Signal",
          "expressions": {
            "$id": "25",
            "Literal": "ValidationSmsSignal"
          }
        }
      ],
      "propertyStorageProviders": {}
    },
    {
      "$id": "26",
      "activityId": "09e515ff-dbae-498e-bc2d-3946a93198e6",
      "type": "CustomSendTrigger",
      "name": "SendMail",
      "displayName": "SendMail",
      "description": "SendMail",
      "persistWorkflow": false,
      "loadWorkflowContext": false,
      "saveWorkflowContext": false,
      "properties": [],
      "propertyStorageProviders": {}
    },
    {
      "$id": "27",
      "activityId": "36eae5c8-bb91-4be1-9ab5-a049f3f18803",
      "type": "CustomSendTrigger",
      "name": "SendSms",
      "displayName": "SendSms",
      "description": "SendSms",
      "persistWorkflow": false,
      "loadWorkflowContext": false,
      "saveWorkflowContext": false,
      "properties": [],
      "propertyStorageProviders": {}
    }
  ],
  "connections": [
    {
      "$id": "28",
      "sourceActivityId": "1c871c36-64e3-4890-9b3d-5a749c0d6d8e",
      "targetActivityId": "191f0857-9f14-46b3-aaef-cb5f2d2a1735",
      "outcome": "Done"
    },
    {
      "$id": "29",
      "sourceActivityId": "191f0857-9f14-46b3-aaef-cb5f2d2a1735",
      "targetActivityId": "1e574748-7698-4106-b0ed-26cb5eefe402",
      "outcome": "Done"
    },
    {
      "$id": "30",
      "sourceActivityId": "d06ae00f-2472-4b68-b757-47ebf92ad3d9",
      "targetActivityId": "949fe4bf-1261-440b-a836-137a4f7ebd2a",
      "outcome": "Done"
    },
    {
      "$id": "31",
      "sourceActivityId": "b63178a5-d02c-4f0a-b19e-e7d42969277b",
      "targetActivityId": "e5656747-01e2-4a6f-a9f5-f36fd9a2bfed",
      "outcome": "Done"
    },
    {
      "$id": "32",
      "sourceActivityId": "1e574748-7698-4106-b0ed-26cb5eefe402",
      "targetActivityId": "09e515ff-dbae-498e-bc2d-3946a93198e6",
      "outcome": "ValidateUsingEmail"
    },
    {
      "$id": "33",
      "sourceActivityId": "09e515ff-dbae-498e-bc2d-3946a93198e6",
      "targetActivityId": "d06ae00f-2472-4b68-b757-47ebf92ad3d9",
      "outcome": "Done"
    },
    {
      "$id": "34",
      "sourceActivityId": "1e574748-7698-4106-b0ed-26cb5eefe402",
      "targetActivityId": "36eae5c8-bb91-4be1-9ab5-a049f3f18803",
      "outcome": "ValidateUsingOtp"
    },
    {
      "$id": "35",
      "sourceActivityId": "36eae5c8-bb91-4be1-9ab5-a049f3f18803",
      "targetActivityId": "b63178a5-d02c-4f0a-b19e-e7d42969277b",
      "outcome": "Done"
    }
  ],
  "id": "89967d7fa6b1462398dc03fb4254f7b7"
}

**Comment by vabs212** [comment]
![image](https://user-images.githubusercontent.com/66916559/130795039-1b04eedc-ca6c-4149-9af4-4662e9d48c00.png)

**Reply by sfmskywalker** [reply]
If you are writing your own activity that acts as a trigger, you will also need to implement a bookmark provider for said trigger. More information about implementing custom blocking activities can be found [here](https://elsa-workflows.github.io/elsa-core/docs/next/guides/guides-blocking-activities).

Is there a specific reason you are writing your own activities that respond to a signal, rather than using the built-in **Signal Received** activity?

**Reply by vabs212** [reply]
Actually we are trying to implement custom dashboard type where user does not need to input Anything .All input acivities will be done in backend. only customer needs to design its own journey and that will be implemented. we do not need to take input in dashboard >all input sort will be done in API call itself.
